torch/distributed/fsdp/_common_utils.py

```diff
@@ -57,14 +57,31 @@ def __init__(self) -> None:
         self._optim_state_dict_config: OptimStateDictConfig = FullOptimStateDictConfig()
         self._is_root: Optional[bool] = None
         self._handles: List[flat_param_file.FlatParamHandle] = []
-        # All FSDP states in the root's tree for the root; `None` for non-root
-        self._fsdp_states: Optional[_FSDPState] = None
         self._fully_sharded_module_to_handles: Dict[
             nn.Module, flat_param_file.FlatParamHandle
         ] = {}
         self.compute_device = torch.device("cuda", torch.cuda.current_device())
 
 
+class _RootFSDPState(_FSDPState):
+    """
+    This class shows the attributes defined only for the root FSDP state (i.e.
+    ones with ``_is_root=True``.) We do not instantiate this class. Instead, we
+    may use the idiom ``state = cast(_RootFSDPState, state)``, where ``state``
+    is an ``_FSDPState`` instance, for improved type checking.
+
+    TODO: Since most FSDP functions have the ``@no_type_check`` decorator, this
+    separation may not have much effect until we remove those.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        # Save these static lists to avoid the repeated tree traversals
+        self._all_fsdp_states: List[_FSDPState] = []
+        self._all_handles: List[flat_param_file.FlatParamHandle] = []
```

Comment:
I feel like it's a little bit convoluted. If `_RootFSDPState` is only for typing purpose and is never instantiated, then the declaration of `_all_fsdp_states` and `_all_handles` are confusing. Basically, this mixes the duck typing with static typing -- we dynamically assign `_all_fsdp_states` to a `_FSDPState` instance and then cast it to `_RootFSDPState` to fulfill the static typing requirement. 

What is the issue of declaring `_all_fsdp_states` and `_all_handles` in `_FSDPState` and leave them to be empty for non-root state?

Reply:
Thanks @fegin  for the insight!

What you said makes sense to me. I think leaving `_all_fsdp_states` and `_all_handles` in `_FSDPState` empty for non-root states sounds good to me.

I was originally thinking that in the existing way, we could still get attribute-not-defined errors if we accidentally try to access for non-root states. (For example, this is the existing approach for `_post_backward_callback_queued`, where in @zhaojuanmao 's original implementation, she wanted that to erroring so that we can catch accidental usages.)

Reply:
Is it worthwhile to make `_all_fsdp_states` and `_all_handles` into properties that check `_is_root == True` upon access?

Reply:
Update: I got rid of `_RootFSDPState`.